ambari-agent/src/main/python/ambari_agent/AmbariAgent.py

```diff
@@ -16,46 +16,85 @@
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
-'''
+"""
 
 import os
 import sys
 from ambari_commons import subprocess32
-import signal
 
 AGENT_AUTO_RESTART_EXIT_CODE = 77
 
-if os.environ.has_key("PYTHON_BIN"):
-  AGENT_SCRIPT = os.path.join(os.environ["PYTHON_BIN"],"site-packages/ambari_agent/main.py")
+
+def get_logger():
+  import logging
+
+  _logger = logging.getLogger(__name__)
+  _logger.handlers = []
+
+  # pattern used at ambari-agent#start to output messages from script. DO NOT change it without changing grep regex
+  formatter = logging.Formatter("%(levelname)s: %(message)s")
+  handler = logging.StreamHandler()
+  handler.setLevel(logging.WARNING)
+  handler.setFormatter(formatter)
+
+  _logger.addHandler(handler)
+
+  return _logger
+
+
+logger = get_logger()
+
+
+if "PYTHON_BIN" in os.environ:
+  AGENT_SCRIPT = os.path.join(os.environ["PYTHON_BIN"], "site-packages/ambari_agent/main.py")
 else:
   AGENT_SCRIPT = "/usr/lib/ambari-agent/lib/ambari_agent/main.py"
-if os.environ.has_key("AMBARI_PID_DIR"):
-  AGENT_PID_FILE = os.path.join(os.environ["AMBARI_PID_DIR"],"ambari-agent.pid")
+
+if "AMBARI_PID_DIR" in os.environ:
+  AGENT_PID_FILE = os.path.join(os.environ["AMBARI_PID_DIR"], "ambari-agent.pid")
 else:
   AGENT_PID_FILE = "/var/run/ambari-agent/ambari-agent.pid"
+
 # AGENT_AUTO_RESTART_EXIT_CODE = 77 is exit code which we return when restart_agent() is called
 status = AGENT_AUTO_RESTART_EXIT_CODE
 
+
+def check_native_libs_support():
+  not_loaded_extensions = []
+
+  from ambari_simplejson import c_extension
+  if not c_extension.is_loaded():
+    not_loaded_extensions.append("simplejson")
+
+  if subprocess32._posixsubprocess is None:
+    not_loaded_extensions.append("subprocess32")
+
+  if not_loaded_extensions:
+    logger.warning("Some native extensions not available for module(s): {}, it may affect execution performance".format(",".join(not_loaded_extensions)))
+
+
 def main():
   global status
 
-  if (os.environ.has_key("PYTHON")):
-    PYTHON = os.environ["PYTHON"]
+  if "PYTHON" in os.environ:
+    python = os.environ["PYTHON"]
   else:
     print("Key 'PYTHON' is not defined in environment variables")
     sys.exit(1)
 
   args = list(sys.argv)
   del args[0]
 
-  mergedArgs = [PYTHON, AGENT_SCRIPT] + args
+  merged_args = [python, AGENT_SCRIPT] + args
 
   while status == AGENT_AUTO_RESTART_EXIT_CODE:
-    mainProcess = subprocess32.Popen(mergedArgs)
-    mainProcess.communicate()
-    status = mainProcess.returncode
+    check_native_libs_support()
```

Comment:
Is check for native lib support needed in each iteration?

Reply:
i have supposed so, as environment could change between "soft" restart with long time delay between those restarts.